ulauncher/VERSION

```diff
@@ -0,0 +1 @@
+6.0.0-dev0
```

Comment:
So when building a version this file needs to first be updated, and presumably checked in, and then it can be tagged? Does it need to get updated back to a dev version immediately after the release is complete, or is a new release branch created first to change the version to? Do we also then need to be sure the exact right version is also passed in to the tar.gz build script. What happens if the version passed in doesn't match the version in the file?

Reply:
> So when building a version this file needs to first be updated, and presumably checked in, and then it can be tagged?

I would do it like that, yes. I don't think it's much to handle, but it's easy to wrap a script around that. The git tag isn't connected to anything. It's just good practice. You could reverse the logic and make the git tag the single source of truth and skip VERSION, but then you can only use setup.py with git, so this really is the easiest way 

> Does it need to get updated back to a dev version immediately after the release is complete

I don't plan on using any of gornostal's scripts for making github releases. They're written for his preferences and are hard coded to work with his launchpad keys and AUR keys for example. To make a release I will just update VERSION, produce the tarball, commit and tag the release and push the release with the tag. I prefer to make the github release and uploading the tarball to it manually. The first github release will be a pre-release and won't have any distro support. Just tarball. Then I hope to get there gradually, but I am looking at using https://openbuildservice.org/ instead of all the distro scripts where you need to run multiple docker containers etc.

Reply:
So I think the two general approaches are:

###  The current procedure: 

* Decide to release v1.0.0
* Checkout master
* Tag master with v1.0.0
* Run ./ul build-targz 1.0.0  (repeat for rpms, etc)
* Upload packages

### In the new scheme: 

* Decide to release v1.0.0
* Checkout master
* Update VERSION to be 1.0.0
* Commit
* Tag with v1.0.0
* Run ./ul build-targz 1.0.0  (repeat for rpms, etc) (manually make sure the passed in value matches the value in VERSION)
* Upload packages
* Remember to go back and update VERSION to be a dev version (1.0.1dev?)
* Commit

It just seems odd to me to have the version hard-coded into the source code repository, and seems to add otherwise unneeded manual steps, which if missed could cause confusion. If changes were made to master before the VERSION file was updated to a "dev" version, then it there would be multiple change sets that all have the same release listed. Presumably the git tag is the more authoritative source, and could be used to resolve the mismatch, but it just illustrates why it might not be ideal to violate DRY principle in this case.

Reply:
Our priorities and ideas of good code, priorities and DRY couldn't be more different, so I think it's pointless to try to collaborate. I will keep pushing this repo in the direction I had in mind, and I guess you can wait for gornostal to return.

Reply:
It doesn't seem pointless, these are just big changes and I'm just trying to help think through the different choices in a deliberate way. I'm also trying to bring along some of my prior experiences. I do probably tend to dive particularly deep with thinking through changes, and I recognize I'm frequently a very...er...detailed.... reviewer. I was pointing out this embedded version issue, because I have directly experienced the forgotten version change issue that I pointed out, and have seen my colleagues tripped up with the same. It's obviously not a huge thing, I just wanted to point out some challenges .

Reply:
Maybe I overreacted, but I found this response to be very hostile/passive aggressive and not at all collaborative or honest.

Node, Ruby gems, Rust, or and probably most other languages with standardized package managers stores the package version in the source, and the [Python Packaging Authority](https://www.pypa.io/en/latest/) suggests the solution in this PR as the recommendation 4 on [their list](https://packaging.python.org/guides/single-sourcing-package-version/) "An advantage with this technique is that it’s not specific to Python. Any tool can read the version." This why I did it this way, so scripts can read from **and write to** the file. The latter of course means there's no additional step at all, because it can all be scripted.

I also already explained this choice: "We could use git tags instead of this file. This would require that everyone running the app as devs without installing first do this with git, so I suggest to keep the file (but use git tags for every release anyway)."

It's actually very easy to make this fork use the git tag again by changing just a few lines of code, but I don't think the benefit outweighs the cost, or that there really are costs to having the version file.

Downloading the repo as a zip for example does not build in upstream Ulauncher. If you run the app directly with `./bin/ulauncher` the version also doesn't work in upstream Ulauncher. It will show `%VERSION%`. If someone forks and builds but doesn't pull in the tags, then it will show the same version as when the fork was first made. This PR fixes those issues, and it removes the extremely dirty solution of setup.py patching itself, and the ulauncher config file. It also removes a lot code (hundreds of lines, but some of those was an svg icon) and is way more simple and DRY than the previous solution, just not the release flow, because that's not the priority. It's already broken and hard coded to work with gornostals repos, identity and package names, so we shouldn't use that.

There's no need to bump the version twice per release. No one using those languages does that, and it's actually misleading if you do. Semver has different meaning for each "part" of the version string, and you don't know which one you should bump until you release, but the version string is fully intended for releases. So what it says between releases doesn't matter anyway.

You also say you have to "manually make sure the passed in value matches the value in VERSION" with this PR, but the inconsistency was already there, except instead of "value in VERSION" it was "value in the git tag". I didn't introduce it in this PR and hence it was just status quo. I already answered that as well.

Reply:
thanks for taking the time for the detailed response. I think I see much better where you are coming from with this and I've learned more than a knew before! I suppose I've always thought of the versions of releases as being sorta distinct snapshots of the code that you put a label on at some arbitrary point in the stream. However, I do realize that is a perspective and not some inherit truth.

Reply:
It seems I did over-react then. Sorry about that, and thank you for your patience with me.